jetty-util/src/main/java/org/eclipse/jetty/util/thread/SerializedExecutor.java

```diff
@@ -34,65 +35,271 @@
 public class SerializedExecutor implements Executor
 {
     private final AtomicReference<Link> _tail = new AtomicReference<>();
+    private final Executor _executor;
+
+    public SerializedExecutor()
+    {
+        this(Runnable::run);
+    }
+
+    public SerializedExecutor(Executor executor)
+    {
+        _executor = executor;
+    }
 
+    private Link newLink(Runnable task)
+    {
+        switch (Invocable.getInvocationType(task))
+        {
+            case EITHER:
+                return new EitherLink(task);
+            case BLOCKING:
+                return new BlockingLink(task);
+            case NON_BLOCKING:
+                return new NonBlockingLink(task);
+            default:
+                throw new IllegalStateException();
+        }
+    }
+
+    /**
+     * Arrange for a task to be invoked, mutually excluded from other tasks.
+     * @param task The task to invoke
+     * @return A Runnable that must be called to invoke the passed task and possibly other tasks. Null if the
+     *         task will be invoked by another caller.
+     */
+    public Runnable offer(Runnable task)
+    {
+        if (task == null)
+            return null;
+        Link link = newLink(task);
+        Link penultimate = _tail.getAndSet(link);
+        if (penultimate == null)
+            return link;
+        penultimate._next.lazySet(link);
+        return null;
+    }
+
+    /**
+     * Arrange for tasks to be invoked, mutually excluded from other tasks.
+     * @param tasks The tasks to invoke
+     * @return A Runnable that must be called to invoke the passed tasks and possibly other tasks. Null if the
+     *         tasks will be invoked by another caller.
+     */
+    public Runnable offer(Runnable... tasks)
+    {
+        Runnable runnable = null;
+        for (Runnable task : tasks)
+        {
+            if (runnable == null)
+                runnable = offer(task);
+            else
+                offer(task);
+        }
+        return runnable;
+    }
+
+    /**
+     * Arrange for a task to be executed, mutually excluded from other tasks.
+     * This is equivalent to executing any {@link Runnable} returned from {@link #offer(Runnable)}
+     * @param task The task to invoke
+     */
     @Override
     public void execute(Runnable task)
```

Comment:
I find the `Runnable offer(Runnable)` vs `void execute(Runnable)` contract confusing.

I think the reason to have a `Executor` implementation that also have this extra `offer` contract that does not queue the task should be better documented.

I'm starting to wonder if the whole `Executor` / `ExecutionStrategy` / `Invocable` mechanism shouldn't be refreshed?

Reply:
I'm definitely wondering what this serializable executor is worth it for the handler arch.  The actions are all serialized by a lock on HttpChannel now, so I'm not sure it is a big problem if multiple callbacks run at the same time.  So I'm dubious that this class is really needed.

However, assuming that it is.... which is the name you find most confusing?    `offer` or `execute`.   Personally I think `execute` is OK as this is just an executor and so externally that is the right contract and internally the mutual exclusion is just a facet of the particular executor.    @sbordet suggested the `offer` name and I initially thought it was good... but now I can see that it is not exactly right, as the executor doesn't have the option to refuse the offer. The semantic is that the task is being submitted to be run and that is done either by another thread or by the runnable that is returned.  Would `submit` be better?  or just javadoc the existing names?

Reply:
What I find confusing is the `offer` contract: queue a task which may or may not get executed as a response to this call and optionally return you a `Runnable` that you have to get executed externally is _extremely_ odd for an `Executor`.

I feel like this contract should live in some class that does not implement `Executor` but does delegate to one.

Reply:
It actually started off that way, without any reference to an `Executor` and was called the `SerialInvoker`, but then there was need for a convenience method that did the execute... so it was added... and then it looked like and executor... which was pretty much the already existing `SerialExecutor` class... so I merged them.

So I agree it's a little it strange.   But can we first ask the big question..... what is it doing for us?   Do we really need it at all?     The original idea was to serialize all the events arriving on `HttpChannel` into a queue, that would not need locks because only a single thread was ever executing the queue (protected by whatever mutex is in this class or equivalent).   But that idea didn't work out because... shit I've forgotten why that didn't work out... we need to write that down.  Anyway, that didn't work out and now we lock on HttpChannel and then serialize the consequences of events.    But do we really need to serialize them?  They can't do any actions significant themselves without first grabbing the lock, so we have a lock and a queue and an lock in the mutex for the queue.... maybe we are the worst of both worlds?

Anyway, I'd like to commit in it's current form and I will remember what the issue was with the fully serialized approach and write that up in another issue for posterity.  We can then consider just removing this class all together (although it is used in one other place).   If we don't remove it, then perhaps rename it and remove the contained executor?

Reply:
@lorban actually, I thought about this some more and serializing does do an important job for this architecture: it prevents infinite recursion of a callback calling a method that calls a callback etc.

So for now, I have split out `SerializedInvoker` from `SerializedExecutor` (although the former name is wrong).   I have also removed the anti pattern of pretending we are a scheduler.  A serialized runnable is always a BLOCKING runnable as a blocking task may be queued at any time.